src/worker/jobs/downloads/update_metadata.rs

```diff
@@ -17,24 +18,52 @@ impl BackgroundJob for UpdateDownloads {
     async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
         spawn_blocking(move || {
             let mut conn = env.connection_pool.get()?;
-            update(&mut conn)?;
+            update(&mut conn, env.config.batch_update_downloads)?;
             Ok(())
         })
         .await
     }
 }
 
-fn update(conn: &mut PgConnection) -> QueryResult<()> {
+fn update(conn: &mut PgConnection, use_batch_update: bool) -> QueryResult<()> {
     use diesel::dsl::now;
     use diesel::select;
 
-    let rows = version_downloads::table
-        .filter(version_downloads::processed.eq(false))
-        .filter(version_downloads::downloads.ne(version_downloads::counted))
-        .load(conn)?;
-
-    info!(rows = rows.len(), "Updating versions");
-    collect(conn, &rows)?;
+    if use_batch_update {
+        info!("Updating versions…");
+
+        // After 45 minutes, we stop the batch updating process to avoid
+        // triggering the long-running job alert. 15 minutes later, the job
+        // will be started again by our cron service anyway.
+        const TIME_LIMIT: Duration = Duration::from_secs(45 * 60);
+
+        // We update the `downloads` columns in batches to a) avoid the
+        // back and forth between the application and the database and b) avoid
+        // holding locks on any of the involved tables for too long.
+        const BATCH_SIZE: i64 = 5_000;
```

Comment:
Empirically, my testing of this suggests that there aren't any locks held during the big update query. Assuming that's generally true, would we want to batch at all?

Reply:
I think I prefer the batching so that we get regular status updates in the logs. If we happen to have 1M rows to update and then the job just doesn't say anything for 2 hours that would seem bad to me.